staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers_test.go

```diff
@@ -77,7 +78,7 @@ func TestLabelSelectorAsSelector(t *testing.T) {
 		if err != nil && !tc.expectErr {
 			t.Errorf("[%v]did not expect error but got: %v", i, err)
 		}
-		if !reflect.DeepEqual(out, tc.out) {
```

Comment:
this doesn't look right... what in LabelSelectorAsSelector was not preserving the existing data?

Reply:
LabelSelectorAsSelector preserves the existing data, it's just the testcase that expect the input data being ordered.

For an input like `"qux", "norf"`:
https://github.com/kubernetes/kubernetes/blob/9d4b574699dfe198d930a6c3282d188581b59f8b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers_test.go#L29-L33

It's expected to be `"baz in (norf,qux),foo=bar"`:

https://github.com/kubernetes/kubernetes/blob/9d4b574699dfe198d930a6c3282d188581b59f8b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers_test.go#L56-L59

So as I commented in https://github.com/kubernetes/kubernetes/pull/66480#issuecomment-407844550, it makes sense to change the testcase expected ouput?

> changed to call String() explicitly. (**or, change the "expected" values to be same order as input**)

Reply:
I see... we could do both...
* have one test case that changes the expected output: `out: mustParse("baz in (qux,norf),foo=bar")`
* have one test case that makes sure `out.String()` and `tc.out.String()` match, regardless of the order of `tc.out`

Reply:
How about this: keep input data as is, and for each test, separate semantics check and string check:

- for input, add semantics check to ensure it's not modifed
- for output, compare to expectedOutput using string check

```diff
        for i, tc := range tc {
+               inCopy := tc.in.DeepCopy()
                out, err := LabelSelectorAsSelector(tc.in)
+               // after calling LabelSelectorAsSelector, tc.in shouldn't be modified
+               if !reflect.DeepEqual(inCopy, tc.in) {
+                       t.Errorf("[%v]expected:\n\t%#v\nbut got:\n\t%#v", i, inCopy, tc.in)
+               }
                if err == nil && tc.expectErr {
                        t.Errorf("[%v]expected error but got none.", i)
                }
                if err != nil && !tc.expectErr {
                        t.Errorf("[%v]did not expect error but got: %v", i, err)
                }
-               if !reflect.DeepEqual(fmt.Sprint(out), fmt.Sprint(tc.out)) {
-                       t.Errorf("[%v]expected:\n\t%+v\nbut got:\n\t%+v", i, tc.out, out)
+               // fmt.Sprint() over String() as nil.String() will panic
+               if fmt.Sprint(out) != fmt.Sprint(tc.out) {
+                       t.Errorf("[%v]expected:\n\t%s\nbut got:\n\t%s", i, fmt.Sprint(tc.out), fmt.Sprint(out))
                }
        }
 }
```